New rest endpoints for Stream

phillip-kruger opened the pull request:
![image](https://user-images.githubusercontent.com/6836179/124127042-8f9a0380-da7b-11eb-8bb6-fa5befcfa7f7.png)

phillip-kruger commented:
@ia3andy please review (I can not add reviewers)

ia3andy commented:
We should add a flag for the community extensions or hide them I am not sure, but we can't merge it without something. @maxandersen any thought?

ia3andy commented:
@phillip-kruger maybe we should update the health check to use the new PlatformService? Also we should make sure we don't break the catalog if an update fails for some reason..

phillip-kruger commented:
@ia3andy I added the flag to filter platform only (default) to the API endpoint. It's now fully backward compatible. Please review. Thanks :)

phillip-kruger commented:
![image](https://user-images.githubusercontent.com/6836179/124584786-ee34f800-de54-11eb-9776-207111c5710b.png)

Ok @ia3andy, health check is changed to now use `PlatformService` and I also added some useful meta data to the health response. I also changed the refresh to keep the old catalog if the call failed.

Please review :)

ia3andy reviewed (approved):
LGTM!

Looking good :)

is the legacy ExtensionCatalog class still used?

phillip-kruger commented:
Not sure ? I only changed the `extensions` endpoints, nothing else. So i have not touched the Project create part. That might need some update too, not sure.

aloubyansky commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,105 @@
+
+    val extensionCatalog: List<CodeQuarkusExtension>?
+        get() {
+            val defaultPlatformKey = platformCatalog!!.recommendedPlatform.platformKey
```
Wouldn't using a local copy of the platformCatalog be safer, e.g.
````
val pc = platformCatalog;
val defaultPlatformKey = pc!!.recommendedPlatform.platformKey
...
````
in case the platformCatalog gets updated in middle?

ia3andy replied:
Yes we should be really careful about the state and concurrency now that it gets refreshed, also maybe it would be safer to use `AtomicReference` for mutable state fields?

phillip-kruger replied:
This is done (not yet committed)

phillip-kruger commented:
@ia3andy please have a look. Create project now also using Platform service. We probably need another endpoint for create that allows passing in the stream (as with the get) but for the default case this should be ok.

phillip-kruger commented:
As we change the UI to allow stream selection we can update the API as needed ?

ia3andy commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,130 @@
+
+    private val catalogResolver = QuarkusProjectHelper.getCatalogResolver()
+    var streamCatalogMap: MutableMap<String, PlatformInfo> = HashMap()
+    var platformCatalog: AtomicReference<PlatformCatalog>? = null
```
```suggestion
    var platformCatalog: AtomicReference<PlatformCatalog> = AtomicReference()
```

ia3andy commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,130 @@
+    @Scheduled(cron = "{reload.catalogs.cron.expr}")
+    fun reloadCatalogs() {
+        try {
+            platformCatalog = AtomicReference(catalogResolver.resolvePlatformCatalog())
```
```suggestion
            platformCatalog.set(catalogResolver.resolvePlatformCatalog())
```

ia3andy commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,130 @@
+
+    val platformInfo: PlatformInfo?
+        get() {
+            val pc = platformCatalog?.get()
```
```suggestion
            val pc = platformCatalog.get()
```

ia3andy commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,130 @@
+    val platformInfo: PlatformInfo?
+        get() {
+            val pc = platformCatalog?.get()
+            val defaultPlatformKey = pc!!.recommendedPlatform.platformKey
```
what if `pc` is null?

ia3andy commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,130 @@
+
+    val extensionCatalog: List<CodeQuarkusExtension>?
+        get() {
+            val pc = platformCatalog?.get()
```
```suggestion
            val pc = platformCatalog.get()
```

ia3andy commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,130 @@
+        }
+
+    fun getDefaultStreamKey(): String {
+        val pc = platformCatalog?.get()
```
```suggestion
        val pc = platformCatalog.get()
```

ia3andy commented on src/main/kotlin/io/quarkus/code/service/PlatformService.kt:
```diff
@@ -0,0 +1,130 @@
+    @Throws(RegistryResolutionException::class)
+    private fun populateExtensionCatalogMaps() {
+        var updatedStreamCatalogMap: MutableMap<String, PlatformInfo> = HashMap()
+        val platforms = platformCatalog?.get()!!.platforms
```
```suggestion
        val platforms = platformCatalog.get()!!.platforms
```

ia3andy reviewed (changes requested):
Looking good, I added a few comments.

I will give it a try after the devtools meeting..
